beacon-chain/db/state.go

```diff
@@ -316,6 +321,80 @@ func (db *BeaconDB) HistoricalStateFromSlot(ctx context.Context, slot uint64) (*
 	return beaconState, err
 }
 
+// ValidatorRegistry fetches the current validator registry stored in state.
+func (db *BeaconDB) ValidatorRegistry(ctx context.Context) ([]*pb.Validator, error) {
+	ctx, span := trace.StartSpan(ctx, "BeaconDB.ValidatorRegistry")
+	defer span.End()
+
+	db.stateLock.RLock()
+	defer db.stateLock.RUnlock()
+
+	// Return in-memory cached state, if available.
+	if db.currentState != nil {
+		_, span := trace.StartSpan(ctx, "proto.Clone.ValidatorRegistry")
+		defer span.End()
+		newRegistry := make([]*pb.Validator, len(db.currentState.ValidatorRegistry))
+		for i, v := range db.currentState.ValidatorRegistry {
+			newRegistry[i] = proto.Clone(v).(*pb.Validator)
+		}
+		return newRegistry, nil
+	}
+
+	var beaconState *pb.BeaconState
+	err := db.view(func(tx *bolt.Tx) error {
+		chainInfo := tx.Bucket(chainInfoBucket)
+		enc := chainInfo.Get(stateLookupKey)
+		if enc == nil {
+			return nil
+		}
+
+		var err error
+		beaconState, err = createState(enc)
+		if beaconState != nil && beaconState.Slot > db.highestBlockSlot {
```

Comment:
should we save `db.currentState` here?

Reply:
hmm, why would we save `db.currentState` ?

Reply:
So we can use cached state next time?

Reply:
this function has to do with just allowing a smaller copy for validator registry and balances, without the need to clone the state again. We only update current state in `SaveState` since thats when we determine our chainhead. It doesnt make sense to modify current state anywhere else

Reply:
I see. It'll prob be dangerous to do so anyway. Everything else looks good to me